### DISCUSSION TITLE
[Feature Request] Render Scaling Support

### DISCUSSION THREAD

**Discussion by lennondotw** [ppy/osu#25208] [Ideas]
When playing osu! on a high resolution screen, we don't have an easy and reliable way to modify the resolution.

Many tablets and laptops (or monitors) are equipped with very high resolutions (4K or even higher). Running osu! in its native resolution on these devices is not necessary and will seriously affect the performance, getting a very low frame rate.

We can only avoid this by running osu! in windowed mode, or by using some tricky methods to modify the system resolution. However, these methods are not always easy and reliable, for example you can't run the game windowed on some Android devices, and modifying the resolution of macOS is not that easy.

Also, I don't think modifying the system resolution to improve performance is a good way to go, it's too "heavy". Some monitors may have to "black screen" for many seconds before you can switch to the new resolution. It's also not a great experience when you're multitasking and switching resolutions back and forth between games and other apps or virtual desktops.

<img width="500" src="https://github.com/ppy/osu/assets/26853900/16e04f6d-b5b6-4113-9a23-cee91b4e5761" />

_(display scaling in macOS)_

Also, if display scaling is turned on in macOS, the game may actually be rendered at a higher resolution on a 3840 x 2160 display. In my case, the game actually renders at 5120 x 2880 (the in-game screenshots illustrate this). Unnecessary waste of performance!

<img width="500" src="https://github.com/ppy/osu/assets/26853900/35dfa0d9-2187-4d7a-969a-7786325a1eb0" />

_(in-game screenshot using F12)_

If we can adjust the rendering resolution in osu! (e.g. by selecting render at 50%), then we could render at 1920 x 1080 and then display fullscreen on a 3840 x 2160 screen. This would be very useful!

**Comment by lennondotw** [comment]
Related discussion: https://github.com/ppy/osu/discussions/20698

Related issue: https://github.com/ppy/osu/issues/10716

**Comment by lennondotw** [comment]
I'm playing osu! lazer on a MacBook Air with an M1 chip hooked up to a 4K@60 display and the framerate is very low. But I think the hardware should be fine because osu! runs very well in windowed mode (rendering resolution is around 2K) and it only uses 50% of the GPU and less than 10% of the CPU.

Also, I have the same issue on my Samsung Galaxy Tab S8+ (Android tablet). It's resolution is 2800 x 1752. the game is very laggy when running in full screen, but runs well in windowed mode.

**Reply by peppy** [reply]
> In my case, the game actually renders at 5120 x 2880 (the in-game screenshots illustrate this). Unnecessary waste of performance!

Wait until you realise the whole operating system is doing this. And that reducing the render scale will cause double artifacting.

**Reply by lennondotw** [reply]
Yes, display scaling is macOS is system level and all apps are rendering at a higher resolution. But this will have a much more noticeable performance impact on a game-like application, so we need a render scaling option even more don't we?

**Reply by peppy** [reply]
I'd hope not. It will add new artifacting.

A better solution may be setting the native resolution lower. Not sure how good the support for that is on macOS these days (we used to allow if via a legacy API but dunno anymore).
